payments/db/sql_store.go

```diff
@@ -65,3 +112,404 @@ func NewSQLStore(cfg *SQLStoreConfig, db BatchedSQLQueries,
 
 // A compile-time constraint to ensure SQLStore implements DB.
 var _ DB = (*SQLStore)(nil)
+
+// fetchPaymentWithCompleteData fetches a payment with all its related data
+// including attempts, hops, and custom records from the database.
+func (s *SQLStore) fetchPaymentWithCompleteData(ctx context.Context,
+	db SQLQueries, dbPayment sqlc.PaymentAndIntent) (*MPPayment, error) {
+
+	// The query will only return BOLT 11 payment intents or intents with
+	// no intent type set.
+	paymentIntent := dbPayment.GetPaymentIntent()
+	paymentRequest := paymentIntent.IntentPayload
+
+	// Fetch payment-level first hop custom records.
+	payment := dbPayment.GetPayment()
+	customRecords, err := db.FetchPaymentLevelFirstHopCustomRecords(
+		ctx, payment.ID,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch payment level custom "+
+			"records: %w", err)
+	}
+
+	// Convert to the FirstHopCustomRecords map.
+	var firstHopCustomRecords lnwire.CustomRecords
+	if len(customRecords) > 0 {
+		firstHopCustomRecords = make(lnwire.CustomRecords)
+		for _, record := range customRecords {
+			firstHopCustomRecords[uint64(record.Key)] = record.Value
+		}
+	}
+
+	// Convert the basic payment info.
+	info := dbPaymentToCreationInfo(
+		payment.PaymentIdentifier, payment.AmountMsat,
+		payment.CreatedAt, paymentRequest, firstHopCustomRecords,
+	)
+
+	// Fetch all HTLC attempts for this payment.
+	attempts, err := s.fetchHTLCAttemptsForPayment(
+		ctx, db, payment.ID,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch HTLC attempts: %w",
+			err)
+	}
+
+	// Set the failure reason if present.
+	var failureReason *FailureReason
+	if payment.FailReason.Valid {
+		reason := FailureReason(payment.FailReason.Int32)
+		failureReason = &reason
+	}
+
+	mpPayment := &MPPayment{
+		SequenceNum:   uint64(payment.ID),
+		Info:          info,
+		HTLCs:         attempts,
+		FailureReason: failureReason,
+	}
+
+	// The status and state will be determined by calling
+	// SetState after construction.
+	if err := mpPayment.SetState(); err != nil {
+		return nil, fmt.Errorf("failed to set payment state: %w", err)
+	}
+
+	return mpPayment, nil
+}
+
+// fetchHTLCAttemptsForPayment fetches all HTLC attempts for a payment and
+// uses ExecuteBatchQuery to efficiently fetch hops and custom records.
+func (s *SQLStore) fetchHTLCAttemptsForPayment(ctx context.Context,
```

Comment:
ah, you can probably come back to this at the performance improvement stage like i did for graph stuff, but i think you can improve performance here quite a bit still: currently it is still only doing payment by payment. So it isnt taking full advantage of the batch query helper. You could instead batch fetch the components for a whole range of payment IDs in one go

Reply:
yes going to address it in this PR